subsys/net/lib/wifi_credentials/wifi_credentials.c

```diff
@@ -0,0 +1,444 @@
+	const char *ssid,
+	enum wifi_security_type *security,
+	uint8_t *bssid_buf,
+	char *psk_buf,
+	char *sae_password_buf,
+	uint32_t *flags
+)
+{
+	/* load entry */
+	uint8_t buf[ENTRY_MAX_LEN] = {0};
+	int ret = wifi_credentials_get_by_ssid_personal_raw(ssid, buf, ARRAY_SIZE(buf));
+
+	if (ret) {
+		return ret;
+	}
+
+	/* unpack entry*/
+	struct wifi_credentials_header *header = (void *)buf;
+
+	*security = header->security;
+	*flags = header->flags;
+	if (header->flags & WIFI_CREDENTIALS_FLAG_BSSID) {
+		if (bssid_buf == NULL) {
+			return -EINVAL;
+		}
+		memcpy(bssid_buf, header->bssid, WIFI_MAC_ADDR_LEN);
+	}
+
+	switch (header->security) {
+	case WIFI_SECURITY_TYPE_NONE:
+		break;
+	case WIFI_SECURITY_TYPE_PSK:
+	case WIFI_SECURITY_TYPE_PSK_SHA256:
+		if (psk_buf == NULL) {
+			return -EINVAL;
+		}
+		struct wifi_credentials_psk *header_psk = (void *)buf;
+
+		memcpy(psk_buf, header_psk->psk, WIFI_PSK_MAX_LEN);
+		break;
+	case WIFI_SECURITY_TYPE_SAE:
+		if (psk_buf == NULL && sae_password_buf == NULL) {
+			return -EINVAL;
+		}
+		struct wifi_credentials_sae *header_sae = (void *)buf;
+
+		if (psk_buf != NULL) {
+			memcpy(psk_buf, header_sae->psk, WIFI_PSK_MAX_LEN);
+		}
+		if (sae_password_buf != NULL) {
+			memcpy(sae_password_buf, header_sae->sae_password,
+			       CONFIG_WIFI_CREDENTIALS_SAE_PASSWORD_LENGTH);
+		}
+		break;
+	default:
+		LOG_ERR("data entry corrupted");
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+int wifi_credentials_delete_by_ssid(
+	const char *ssid
+)
+{
+	ssize_t idx = lookup_idx(ssid);
+
+	if (idx == -1) {
+		return 0;
+	}
+	memset(entries[idx].ssid, 0, WIFI_SSID_MAX_LEN);
+
+	sys_slist_find_and_remove(&wifi_credentials_entry_list, &entries[idx].entry);
+
+	return wifi_credentials_delete_entry(idx);
+}
+
+sys_slist_t *wifi_credentials_list_ssids(void)
+{
+	return &wifi_credentials_entry_list;
+}
+
+#if defined(CONFIG_WIFI_CREDENTIALS_BACKEND_SETTINGS)
+#include <zephyr/settings/settings.h>
+
+BUILD_ASSERT(ENTRY_MAX_LEN <= SETTINGS_MAX_VAL_LEN);
+
+/**
+ * @brief WiFi credentials settings handler
+ * @note gets called when settings are loaded
+ *
+ * @param key credential index
+ * @param len length of entry
+ * @param read_cb read callback
+ * @param cb_arg argument for read callback
+ * @return 0 on success, -ERRNO on error
+ */
+static int settings_set(const char *key, size_t len, settings_read_cb read_cb, void *cb_arg)
+{
+	/* key validation */
+	if (!key) {
+		return -EINVAL;
+	}
+	int idx = atoi(key);
+
+	if ((idx == 0 && strcmp(key, "0") != 0) ||
+	     idx >= CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES) {
+		return -EINVAL;
+	}
+	if (len < sizeof(struct wifi_credentials_header)) {
+		return -EINVAL;
+	}
+	uint8_t buf[ENTRY_MAX_LEN];
+	size_t length_read = read_cb(cb_arg, buf, ARRAY_SIZE(buf));
+
+	/* value validation */
+	if (length_read < len) {
+		return -ENODATA;
+	}
+
+	load_entry(idx, buf, length_read);
+	return 0;
+}
+
+int wifi_credentials_pickle_entry(size_t idx, const uint8_t *buf, size_t buf_len)
```

Comment:
Don't prefix with `wifi` since this is a private function. Should also be static. Same goes for delete function.

Reply:
I do both of these things for testing: These functions connect the library to different backends. For the test, I can provide a dummy backend that way. I would like to keep that.

Reply:
I think this should be aligned across NCS, it's a pretty common issue, and we should avoid leaking things out of scope.
We've solved it like this earlier, but that's not necessarily the best solution: https://github.com/nrfconnect/sdk-nrf/blob/main/applications/asset_tracker_v2/src/modules/ui_module.c#L30